File: leetcode/1094-car-pooling/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        vector<int> delta(1001, 0);
        for (auto& trip : trips) {
            delta[trip[1]] += trip[0];
            delta[trip[2]] -= trip[0];
        }

        int passengers = 0;
        for (int change : delta) {
            passengers += change;
            if (passengers > capacity) {
                return false;
            }
        }
        return true;
    }
};

File: leetcode/1094-car-pooling/solution.py
from typing import List


class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        delta = [0] * 1001
        for num, start, end in trips:
            delta[start] += num
            delta[end] -= num

        passengers = 0
        for change in delta:
            passengers += change
            if passengers > capacity:
                return False
        return True
